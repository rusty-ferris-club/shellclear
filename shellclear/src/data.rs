use serde_derive::Deserialize;

use crate::shell::Shell;

pub const ROOT_APP_FOLDER: &str = env!("CARGO_PKG_NAME");

#[derive(Debug)]
pub struct CmdExit {
    pub code: exitcode::ExitCode,
    pub message: Option<String>,
}

#[derive(Debug, Deserialize, Clone)]
pub struct SensitiveCommands {
    #[serde(with = "serde_regex")]
    pub test: regex::Regex,
    pub name: String,
    pub secret_group: u8,
    #[serde(default)]
    pub id: String,
}

#[derive(Debug, Clone)]
pub struct FindingSensitiveCommands {
    pub shell_type: Shell,
    pub sensitive_findings: Vec<SensitiveCommands>,
    pub command: String,
    pub data: String,
    pub secrets: Vec<String>,
}
