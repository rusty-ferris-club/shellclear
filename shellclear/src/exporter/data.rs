use anyhow::Result;
use chrono::{DateTime, NaiveDateTime, TimeZone, Utc};
use lazy_static::lazy_static;
use regex::Regex;

use crate::{
    data::FindingSensitiveCommands,
    shell::{FishHistory, Shell},
};

const DATE_TIME_FORMAT: &str = "%Y-%m-%d %H:%M:%S";
pub const LIMIT_COMMAND: usize = 100;

pub trait Exporter {
    /// export sensitive findings results
    ///
    /// # Errors
    ///
    /// Will return `Err` export has an error
    fn sensitive_data(&self, findings: &[FindingSensitiveCommands]) -> Result<()>;
}

lazy_static! {
    static ref ZSHRC_CAPTURE_COMMAND_TIME: Regex = Regex::new(r"^: ([0-9]+):").unwrap();
}

pub fn chunk(text: &str, size: usize) -> String {
    text.chars()
        .collect::<Vec<char>>()
        .chunks(size)
        .map(|c| c.iter().collect::<String>())
        .collect::<Vec<String>>()
        .join("\r\n")
}

/// Get command time execution
///
/// # Errors
///
/// Will return `Err` when conversion error
pub fn extract_time(finding: &FindingSensitiveCommands) -> Result<String> {
    match finding.shell_type {
        Shell::Zshrc => {
            if let Some(c) = ZSHRC_CAPTURE_COMMAND_TIME.captures(&finding.data) {
                if let Some(timestamp) = c.get(1) {
                    return Ok(format!(
                        "{}",
                        convert_str_timestamp_to_date_time(timestamp.as_str())?
                            .format(DATE_TIME_FORMAT)
                    ));
                }
            };
            Ok(finding.data.clone())
        }
        Shell::Fish => {
            let history: FishHistory = serde_yaml::from_str(&finding.data)?;
            Ok(format!(
                "{}",
                convert_str_timestamp_to_date_time(history.when.as_str())?.format(DATE_TIME_FORMAT)
            ))
        }
        _ => Ok("".to_string()),
    }
}

fn convert_str_timestamp_to_date_time(timestamp: &str) -> Result<DateTime<Utc>> {
    Ok(Utc.from_utc_datetime(&NaiveDateTime::from_timestamp(timestamp.parse::<i64>()?, 0)))
}

#[cfg(test)]
mod test_exporter {
    use std::str;

    use insta::assert_debug_snapshot;

    use super::*;
    use crate::shell::Shell;

    #[test]
    fn can_extract_time_zshrc() {
        let shell_finding = FindingSensitiveCommands {
            shell_type: Shell::Zshrc,
            sensitive_findings: vec![],
            command: "test command".to_string(),
            data: ": 1655110559:0;command data".to_string(),
            secrets: vec![],
        };

        let resp = extract_time(&shell_finding);

        assert_debug_snapshot!(resp);
    }

    #[test]
    fn can_extract_time_fish() {
        let shell_finding = FindingSensitiveCommands {
            shell_type: Shell::Fish,
            sensitive_findings: vec![],
            command: "test command".to_string(),
            data: r#"{ cmd: "export test command", when: "1655110559" }"#.to_string(),
            secrets: vec![],
        };

        let resp = extract_time(&shell_finding);

        assert_debug_snapshot!(resp);
    }
}
