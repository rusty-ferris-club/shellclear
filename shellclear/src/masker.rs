use mask_text::Kind;

use crate::data::Command;

const DEFAULT_PERCENTAGE: u8 = 80;
const DEFAULT_MIN_CHARS: usize = 3;
const DEFAULT_MASK_CHARS: &str = "*";

pub struct Masker {
    percentage: u8,
    min_chars: usize,
    mask_chars: String,
}

impl Masker {
    pub fn new() -> Self {
        Self {
            percentage: DEFAULT_PERCENTAGE,
            min_chars: DEFAULT_MIN_CHARS,
            mask_chars: DEFAULT_MASK_CHARS.to_string(),
        }
    }

    pub fn mask_sensitive_findings(&self, results: &mut [Command]) {
        let commands_with_detections = Masker::get_commands_with_detection(results);

        for command in commands_with_detections {
            let should_mask = command.secrets.iter().any(|secret| {
                let replaced_secret = Kind::Percentage(
                    secret.clone(),
                    self.percentage,
                    self.min_chars,
                    self.mask_chars.clone(),
                )
                .mask();

                // there can still be an false-positive here, if the text is all made up of
                // asterisks
                !command.command.contains(&replaced_secret)
            });

            if !should_mask {
                // If all the secrets are already masked, set the detections to an empty vec,
                // this represents that this command is already masked / no secrets detected
                command.detections = vec![];
                continue;
            }

            // Can be optimized, we can chose to not re-mask the secret, using a map, etc...
            for secret in &command.secrets {
                let replaced_secret = Kind::Percentage(
                    secret.clone(),
                    self.percentage,
                    self.min_chars,
                    self.mask_chars.clone(),
                )
                .mask();

                command.command = command.command.replace(secret, &replaced_secret);
                command.data = command.data.replace(secret, &replaced_secret);
            }
        }
    }

    fn get_commands_with_detection(commands: &mut [Command]) -> Vec<&mut Command> {
        commands
            .iter_mut()
            .filter(|c| !c.detections.is_empty())
            .collect::<Vec<_>>()
    }
}

#[cfg(test)]
mod test_masker {
    use anyhow::Result;
    use insta::assert_debug_snapshot;
    use regex::Regex;

    use crate::{
        data::{Command, Detection},
        masker::Masker,
        shell::Shell::Zshrc,
    };

    #[test]
    fn mask_results() -> Result<()> {
        let mut commands = vec![Command {
            shell_type: Zshrc,
            detections: vec![Detection {
                test: Regex::new("export (MASK_ME)")?,
                name: "mask me mock".to_string(),
                secret_group: 1,
                id: "".to_string(),
            }],
            command: "export MASK_ME".to_string(),
            data: "export MASK_ME".to_string(),
            secrets: vec!["MASK_ME".to_string()],
        }];

        Masker::new().mask_sensitive_findings(commands.as_mut());

        assert_debug_snapshot!(commands);

        Ok(())
    }

    #[test]
    fn remove_already_masked_detection() -> Result<()> {
        let mut commands = [Command {
            shell_type: Zshrc,
            detections: vec![Detection {
                test: Regex::new("export (MASK_ME)")?,
                name: "mask me mock".to_string(),
                secret_group: 1,
                id: "".to_string(),
            }],
            command: "export MA*****".to_string(),
            data: "export MA*****".to_string(),
            secrets: vec!["MASK_ME".to_string()],
        }];

        Masker::new().mask_sensitive_findings(commands.as_mut());

        assert_debug_snapshot!(commands);

        Ok(())
    }
}
