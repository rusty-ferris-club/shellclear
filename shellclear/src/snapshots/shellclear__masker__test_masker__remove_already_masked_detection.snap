---
source: shellclear/src/masker.rs
assertion_line: 123
expression: commands
---
[
    Command {
        shell_type: Zshrc,
        detections: [],
        command: "export MA*****",
        data: "export MA*****",
        secrets: [
            "MASK_ME",
        ],
    },
]
